### PR TITLE
#167727728 integrate facebook and google login

### PR DESCRIPTION
**Pull request by ladaposamuel**
#### What does this PR do?
This PR will enable facebook and google login

#### Description of Task to be completed?
- installed passport library
- create passport strategy
- create facebook and google login routes
- create callback routes
- update user model file with password set to null
- write tests for new endpoints

#### How should this be manually tested?
- CLONE this repo
- open the cloned directory
- run `npm run start:dev`
- access `/auth/google` or `/auth/facebook`

#### What are the relevant pivotal tracker stories?

#167727728

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -157,8 +157,22 @@ const updateUserDetails = async (req, res) => {
   }
 };
 
+const socialAuth = async (req, res) => {
+  try {
+    const {
+      id, firstname, lastname, email
+    } = req.user;
+    const token = authHelper.generateToken({ id });
+    return response(res, 200, 'success', {
```
Comment (tejiri4) [inline src/controllers/userController.js]: Redirect the user to the frontend url with a token in the url either as query

File: src/config/passport.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+const facebookCallback = async (accessToken, refreshToken, profile, done) => {
+  try {
+    const { _json: userDetails } = profile;
+    const { email } = userDetails;
+    let user = await findByEmail(email) || {};
+    if (Object.keys(user).length === 0) {
+      // create user
+      const [firstName, lastName] = userDetails.name.split(' ');
+      user.id = uuidv4();
```
Comment (tejiri4) [inline src/config/passport.js]: Why use uuidv4 here?

File: src/config/passport.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      user.firstName = firstName;
+      user.lastName = lastName;
+      user.email = userDetails.email;
```
Comment (tejiri4) [inline src/config/passport.js]: Email was destructure in line 43

File: src/config/passport.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  callbackURL: `${baseUrl}/api/v1/auth/google/callback`,
+  passReqToCallback: true
+};
+
+const facebookCallback = async (accessToken, refreshToken, profile, done) => {
+  try {
+    const { _json: userDetails } = profile;
+    const { email } = userDetails;
+    let user = await findByEmail(email) || {};
+    if (Object.keys(user).length === 0) {
```
Comment (tejiri4) [inline src/config/passport.js]: Why not use if(user)?

File: src/index.js [hunk 5]
```diff
@@ -35,6 +39,20 @@ app.use(express.urlencoded({ extended: false }));
...
+app.use(
+  session({
+    secret: process.env.SESSION_SECRET,
+    cookie: { maxAge: 60000 },
+    resave: false,
+    saveUninitialized: false
+  })
+);
+app.use(passport.initialize());
+app.use(passport.session());
```
Comment (Mcdavid95) [inline src/index.js]: why use passport session, I thought the facebook and google use OAuth?

File: src/index.js [hunk 6]
```diff
@@ -2,14 +2,18 @@ import express from 'express';
 import trimmer from 'trim-request-body';
 import cors from 'cors';
 import debug from 'debug';
+import methodOverride from 'method-override';
 import swaggerUi from 'swagger-ui-express';
+import passport from 'passport';
+import session from 'express-session';
```
Comment (Mcdavid95) [inline src/index.js]: Why are you using `express-session`, thought we are using jwt already, to make the API stateless

Comment (ladaposamuel) [reply src/index.js]: Ive removed it , thanks for noting

**Review by tejiri4 — APPROVED** [review]
LGTM

**Pull request merged**